File: app/src/main/java/com/elegion/android/util/moxy/strategy/SkipByTagStrategy.java
package com.elegion.android.util.moxy.strategy;

import com.arellomobile.mvp.MvpView;
import com.arellomobile.mvp.viewstate.ViewCommand;
import com.arellomobile.mvp.viewstate.strategy.StateStrategy;

import java.util.Iterator;
import java.util.List;

/**
 * Command will not be put in commands queue. If commands queue contains commands with same tag, then existing commands will be removed.
 *
 * @author mikhail.barannikov on 25.07.2017
 */
public class SkipByTagStrategy implements StateStrategy {
    @Override
    public <T extends MvpView> void beforeApply(List<ViewCommand<T>> currentState, ViewCommand<T> incomingCommand) {
        Iterator<ViewCommand<T>> iterator = currentState.iterator();

        while (iterator.hasNext()) {
            ViewCommand<T> entry = iterator.next();

            if (entry.getTag().equals(incomingCommand.getTag())) {
                iterator.remove();
                break;
            }
        }
    }

    @Override
    public <T extends MvpView> void afterApply(List<ViewCommand<T>> currentState, ViewCommand<T> incomingCommand) {
        // pass
    }
}

File: app/src/main/kotlin/com/elegion/android/bankex/util/moxy/strategy/SkipByTagStrategy.kt
package com.elegion.android.bankex.util.moxy.strategy

import com.arellomobile.mvp.MvpView
import com.arellomobile.mvp.viewstate.ViewCommand
import com.arellomobile.mvp.viewstate.strategy.StateStrategy

/**
 * Command will not be put in commands queue. If commands queue contains commands with same tag, then existing commands will be removed.
 */
class SkipByTagStrategy : StateStrategy {
    override fun <T : MvpView> beforeApply(currentState: MutableList<ViewCommand<T>>, incomingCommand: ViewCommand<T>) {
        val iterator = currentState.iterator()
        while (iterator.hasNext()) {
            val entry = iterator.next()
            if (entry.tag == incomingCommand.tag) {
                iterator.remove()
                break
            }
        }
    }

    override fun <T : MvpView> afterApply(currentState: List<ViewCommand<T>>, incomingCommand: ViewCommand<T>) {
        // pass
    }
}
